httpcore/_utils.py

```diff
@@ -63,3 +64,20 @@ def origin_to_url_string(origin: Origin) -> str:
     scheme, host, explicit_port = origin
     port = f":{explicit_port}" if explicit_port != DEFAULT_PORTS[scheme] else ""
     return f"{scheme.decode('ascii')}://{host.decode('ascii')}{port}"
+
+
+def is_socket_at_eof(sock_fd: int) -> bool:
+    # Duplicate the socket from the file descriptor. (we do this so we always get a
+    # real `socket.socket` object, rather than a library-provided socket-like object,
+    # which may behave differently and/or may not have an implementation for `.recv()`.)
+    sock = socket.fromfd(sock_fd, socket.AF_INET, socket.SOCK_STREAM)
```

Comment:
I guess maybe an important note is that this is TCP-only, although it wouldn't be anything else (yet).

I was looking through some Trio code recently trying to learn about sockets and I saw this code to support different socket types: https://github.com/python-trio/trio/blob/b74f1d85a9c8e23950f49f72bb5fcd28534dfff2/trio/_socket.py#L290-L307

Not sure if we also need to do `sock.detach()`?

Reply:
> Not sure if we also need to do sock.detach()?

Reading more closely and I think the answer is "no". `socket.fromfd` dup's the file descriptor so we do want it to close that.

(Also I'm not really suggesting any change here, just linking to that Trio source out of interest)

Reply:
> although it wouldn't be anything else

Actually it could, if the user is using a Unix Domain Socket… 😕